### ISSUE TITLE
Change handling of tabs.

### ISSUE THREAD

**Issue by sangisos** [goccmack/gocc#35]
As previously discussed in https://github.com/goccmack/gocc/issues/14#issuecomment-254088561 , the handling of tabs as 4 (or any other number) of speces are not satisfactory. I propose to remove special handling of tab characters all togheter, with will not only better conform to other compilators and many text editors, but also simplify any backing up in the source.

**Comment by mewmew** [comment]
Sounds good to me.  Getting rid of the special case makes tabs size 1.

**Comment by awalterschulze** [comment]
Is there any example of how other languages report these syntax errors?

**Comment by awalterschulze** [comment]
I don't like taking examples from the javascript world, but anyway this seems really verbose, but also very unambiguous.

```
/home/substack/projects/node-syntax-error/example/src.js:5
        if (Array.isArray(x) res.push.apply(res, x);
                             ^
ParseError: Unexpected identifier
```

**Comment by awalterschulze** [comment]
But this is just an opinion at the moment.

**Comment by mewmew** [comment]
For an example of carrot style error messages implemented by @sangisos, see https://github.com/mewmew/uc/blob/master/sem/errors/errors.go

**Comment by awalterschulze** [comment]
Ok wow thats awesome.
But maybe we can just remove the color parts, because that adds an extra dependency
```"github.com/mewkiz/pkg/term"```
Or if we think color is really cool, this might be a nice case to start using some govendor -ing

**Comment by mewmew** [comment]
Lets skip colours for now.

**Comment by mewmew** [comment]
The key line for tab size is

```go
srcLine = strings.Replace(srcLine, "\t", " ", -1)
```

Since then, each tab is printed as a space of size 1.

**Comment by awalterschulze** [comment]
Yes but if we don't have the carrot I have to go to my text editor of choice and find the character that is causing the error there.  In my current editor, which has a tab size of 4, the column numbers jump by 4 for each tab.  If the error being reported has a tab size of one, then we have confusion.
So having the carrot type error reporting clears up all possible ambiguities between text editors, because I don't have to look in my text editor to understand the error.

**Comment by mewmew** [comment]
Indeed. I did not express the intention clearly. Carrots should definitely be used, so that error messages in the terminal are self contained.

However, we should print spaces and not tabs in the terminal, because we cannot know the tab size of the terminal. That was my point.

Print carrot error messages in the terminal (or wherever Gocc is invoked from), using spaces for indentation as the tab size cannot be known.

**Comment by awalterschulze** [comment]
awesome I agree.
I was clear on what you meant, but I didn't know if it was clear to everyone.
So I was just extra verbose.

Ok well @sangisos if you agree, then I think you have your design :)

**Comment by mewmew** [comment]
Perfect.

@sangisos Do you wanna prototype this? Remove the special case code for tab handling, and give error handling some love?

**Comment by sangisos** [comment]
Yes. I think the above is ok for most cases, just some concerns and clarifications.

1. Even whitout knowing the tabsize for a terminal, we can just leave all whitespace untouched replacing printables to get the carrot in the right position.

2. Long lines will have the carrot pointing to a lower line if line wrapped:
```c
(?:?) invalid argument
						Example example = example(wrong,
						                          example[example]
						                  );

						                          ^

```
which makes inline element such as bold helpful. Maybe we could use an inline marker which we do not need to change depending on VT-100 complience. Or maybe we cut out the offending part of lines that are over 80 cols.

**Comment by awalterschulze** [comment]
Good point.
Ok so then we might have some long lines like
```c
(?:?) invalid argument
						Example example = example(wrong,
						                          example[example]
						                  );
						                          ^

```
And when we remove the non offending part
```c
(?:?) invalid argument
						Example example = example(wrong,...
						                          ^
```

**Comment by sangisos** [comment]
Yes. I'll try prototyping as soon as i have fixed my linux installation.

**Comment by awalterschulze** [comment]
Sounds good to me :)
